README.md
# saberrs

[![crates.io version badge](https://img.shields.io/crates/v/saberrs.svg)](https://crates.io/crates/saberrs)
[![Documentation](https://docs.rs/saberrs/badge.svg)](https://docs.rs/saberrs)
![License](https://img.shields.io/crates/l/log.svg)
![CI](https://github.com/fparat/saberrs/workflows/Rust/badge.svg)

`saberrs` is a library for interfacing with [Dimension Engineering]
Sabertooth motor driver.

Support:
- Sabertooth 2x32
- Sabertooth 2x60

Full documentation: https://docs.rs/saberrs

## Simple usage

In `Cargo.toml`:

```toml
[dependencies]
saberrs = "0.4"
```

In application code:

```rust
use saberrs::sabertooth2x32::{Sabertooth2x32, PacketSerial};

// Create a handle. This will use "PacketSerial" protocol.
let mut saber = PacketSerial::new("/dev/ttyS0")?;

// Go forward at half-speed (50.0%)
saber.set_drive(0.5)?;
saber.set_turn(0.0)?;

// Request the battery voltage from motor 1.
let vbat : f32 = saber.get_voltage(1)?;

// Stop the motors
saber.stop_motors()?;

```

Other protocol variants can be used:

```rust
use saberrs::sabertooth2x32::{Sabertooth2x32, PacketSerial, PacketType, PlainText};

// "PacketSerial" with specified address and frame protection type (checksum or crc).
let mut saber = PacketSerial::new("/dev/ttyS0")?
    .with_packet_type(PacketType::Checksum)
    .with_address(129);

// "PlainText" protocol
let mut sabertext = PlainText::new("/dev/ttyS1")?;
```

## Features and dependencies

Features:

- `serialport`, enabled by default, for providing default serial IO handlers.

Dependencies:

- [serialport] for the `serialport` feature.
- [log] for emitting logs.

## License

Licensed under either of

* Apache License, Version 2.0
([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
* MIT license
([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

## Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.


## Disclaimer

This library is not affiliated or associated in any way with Dimension Engineering.

All product and company names are trademarks or registered trademarks of
their respective holders. Use of them does not imply any affiliation with or
endorsement by them.

[Dimension Engineering]: https://www.dimensionengineering.com
[serialport]: https://crates.io/crates/serialport
[log]: https://crates.io/crates/log


CHANGELOG.md
# Changelog

## Next release
### Changed
- Update dependency `seriaport` to v4.0
- **Breaking change**: simpler Error type

### Added
- Sabertooth 2x60 support

## v0.3.1 (2020-08-15)
- Fix documentation

## v0.3.0 (2020-08-15)
- **Breaking change**: changed the parameters format of functions taking a
  percentage between -100.0 and +100.0. Now those parameters take a ratio
  between -1.0 and +1.0 instead.

## v0.2.1 (2019-10-13)
- Add "serialport" feature, optional and enabled by default. The external crate
  "serialport" is a dependency only if the feature "serialport" is enabled.
- The `From` implementations for `PacketSerial` and `PlainText` are generic for
  any `SabertoothSerial` object.
- Remove useless "cargo-readme" dependency.
- Add logging using "log" external crate.
- Improve documentation.

## v0.2.0 (2019-08-27)
First release. Support of Sabertooth 2x32.


src/lib.rs
//! `saberrs` is a library for interfacing with [Dimension Engineering]
//! Sabertooth motor driver.
//!
//! Support:
//! - Sabertooth 2x32
//! - Sabertooth 2x60
//!
//! # Simple usage
//!
//! ```rust
//! # use saberrs::Result;
//! use saberrs::sabertooth2x32::{Sabertooth2x32, PacketSerial};
//!
//! # fn example() -> Result<()> {
//! // Create a handle. This will use "PacketSerial" protocol.
//! let mut saber = PacketSerial::new("/dev/ttyS0")?;
//!
//! // Go forward at half-speed (50.0%)
//! saber.set_drive(0.5)?;
//! saber.set_turn(0.0)?;
//!
//! // Request the battery voltage from motor 1.
//! let vbat : f32 = saber.get_voltage(1)?;
//!
//! // Stop the motors
//! saber.stop_motors()?;
//!
//! # Ok(())
//! # }
//! ```
//!
//! Other protocol variants can be used:
//!
//! ```rust
//! use saberrs::sabertooth2x32::{Sabertooth2x32, PacketSerial, PacketType, PlainText};
//! # use saberrs::Result;
//!
//! # fn example() -> Result<()> {
//! // "PacketSerial" with specified address and frame protection type.
//! let mut saber = PacketSerial::new("/dev/ttyS0")?
//!     .with_packet_type(PacketType::Checksum)
//!     .with_address(129);
//!
//! // "PlainText" protocol
//! let mut sabertext = PlainText::new("/dev/ttyS1")?;
//! # Ok(())
//! # }
//! ```
//!
//! # Customizing the IO: the `SabertoothSerial` trait
//!
//! The handles rely on the trait [SabertoothSerial], which abstract the
//! low-level IO communication with the device.
//!
//! By default, the library provides [SabertoothPort] and
//! [SabertoothPortShared]. In most cases the application writer shouldn't need
//! to care about those, but they may be used for applying custom baud rates or
//! timeout values for example.
//!
//! [SabertoothSerial] can be implemented manually for even more customization.
//! For example stubs can be implemented for debugging purpose:
//!
//! ```rust
//! use std::time::Duration;
//! use std::io::{self, Read, Write};
//! use saberrs::{SabertoothSerial};
//!
//! struct SerialStub {
//!     timeout: Duration,
//!     baudrate: u32,
//! }
//!
//! impl SerialStub {
//!     pub fn new() -> Self {
//!         SerialStub {
//!             timeout: Duration::from_millis(100),
//!             baudrate: 9600,
//!         }
//!     }
//! }
//!
//! impl Read for SerialStub {
//!     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
//!         println!("SerialStub.read()");
//!         Ok(0)
//!     }
//! }
//!
//! impl Write for SerialStub {
//!     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
//!         println!("SerialStub.write({:?})", buf);
//!         Ok(buf.len())
//!     }
//!
//!     fn flush(&mut self) -> io::Result<()> { Ok(()) }
//! }
//!
//! impl SabertoothSerial for SerialStub {
//!     fn set_timeout(&mut self, timeout: Duration) -> saberrs::Result<()> {
//!         println!("SerialStub.set_timeout({:?})", timeout);
//!         self.timeout = timeout;
//!         Ok(())
//!     }
//!
//!     fn timeout(&self) -> Duration {
//!         println!("SerialStub.timeout() -> {:?}", self.timeout);
//!         self.timeout
//!     }
//!
//!     fn set_baud_rate(&mut self, baud_rate: u32) -> saberrs::Result<()> {
//!         println!("SerialStub.set_baudrate({})", baud_rate);
//!         self.baudrate = baud_rate;
//!         Ok(())
//!     }
//!
//!     fn baud_rate(&self) -> saberrs::Result<u32> {
//!         println!("SerialStub.baud_rate() -> {}", self.baudrate);
//!         Ok(self.baudrate)
//!     }
//!
//!     fn clear_all(&self) -> saberrs::Result<()> { Ok(()) }
//! }
//! ```
//!
//!
//! # Features and dependencies
//!
//! Features:
//!
//! - `serialport`, enabled by default, allows the usage of the crate
//! [serialport] for providing [SabertoothPort] and [SabertoothPortShared].
//! If this feature is disabled [SabertoothSerial] needs to be implemented
//! manually.
//!
//! Dependencies:
//!
//! - [serialport] for the `serialport` feature.
//! - [log] for emitting logs.
//!
//! # Disclaimer
//!
//! This library is not affiliated or associated in any way with Dimension Engineering.
//!
//! All product and company names are trademarks or registered trademarks of
//! their respective holders. Use of them does not imply any affiliation with or
//! endorsement by them.
//!
//! [Dimension Engineering]: https://www.dimensionengineering.com
//! [Sabertooth 2x32]: https://www.dimensionengineering.com/products/sabertooth2x32
//! [Sabertooth 2x60]: https://www.dimensionengineering.com/products/sabertooth2x60
//! [SabertoothSerial]: trait.SabertoothSerial.html
//! [SabertoothPort]: struct.SabertoothPort.html
//! [SabertoothPortShared]: struct.SabertoothPortShared.html
//! [serialport]: https://crates.io/crates/serialport
//! [log]: https://crates.io/crates/log

pub use error::{Error, Result};
pub use port::SabertoothSerial;

#[cfg(feature = "serialport")]
pub use port::sabertoothport::{SabertoothPort, SabertoothPortShared};

#[macro_use]
mod utils;

mod error;
mod port;

/// Interface for the [Sabertooth 2x32].
///
/// [Sabertooth 2x32]: https://www.dimensionengineering.com/products/sabertooth2x32
pub mod sabertooth2x32;

/// Interface for the [Sabertooth 2x60].
///
/// [Sabertooth 2x60]: https://www.dimensionengineering.com/products/sabertooth2x60
pub mod sabertooth2x60;
